cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -204,6 +180,37 @@ class TypeInferrer {
   int64_t total_count() const { return total_count_; }
 
  protected:
+  Status VisitDType(PyArray_Descr* dtype) {
+    // This is a bit silly.  NumPyDtypeToArrow() infers a Arrow datatype
+    // for us, but we go back to counting abstract type kinds.
```

Comment:
I think it's because there could be a mix of NumPy and non-NumPy things? Is there potentially a better way?

Reply:
I don't know. In Numba we had something called type unification where, for example, `unify(int32, int64) == int64`. It may be costlier than simply incrementing counters, though.

Reply:
Makes sense. We'll need to implement a similar kind of thing here eventually to be able to create unions and so forth. 

I created ARROW-2899 to track